rfcs/rfc0017.md

```diff
@@ -0,0 +1,149 @@
+# built-in functions for checking the original type of a scalar
+
+## Preamble
+
+    Author:  Graham Knop <haarg@haarg.org>
+    Sponsor:
+    ID:      0017
+    Status:  Draft
+
+## Abstract
+
+This RFC proposes adding new builtin functions created_as_string and
+created_as_number for checking the original form of non-scalar references.
```

Comment:
Shouldn't  it be created_as_integer and also created_as_float? Or are we not going to differentiate between NV and IV?

Reply:
It is not yet possible to make that distinction. It is included in the "Future Scope" section though. I think a check for "number" is still valuable.

Reply:
Would it maybe make more sense to implement something like `%flags = scalar_flags($foo)` or `scalar_props($foo)` that returns all the interesting bits you do not normally see, including the things these functions look at, plus some summary keys for `created_as_xxx`? Then document the gotchas and what each thing may or may not mean?

Or are these specific questions really important to answer while the other things contained under the hood of a scalar are useless at the perl-code level?

Reply:
That kind of information is already available via B, but it is hard to interpret correctly. And the specific behavior of the internal flags may change over time. The fact that they can change without it being a backwards compatibility break is in fact why this RFC is viable at all. Making these flags more visible would often be misleading, as they will change based on read operations. The changes in perl 5.36 make it possible to use the flags to accurately check the type, but only for the specific case of string vs number.